crates/metadata-store/src/raft/store.rs

```diff
@@ -580,6 +596,56 @@ impl Active {
             self.on_ready().await?;
             self.update_leadership();
         }
+
+        self.fail_pending_requests();
+
+        let mut storage = self.raw_node.raft.r.raft_log.store;
+        // prevent us from starting as a member again
+        storage.delete_raft_configuration().await?;
+
+        let passive = Passive::new(
+            storage,
+            self.connection_manager,
+            self.request_rx,
+            self.join_cluster_rx,
+            self.metadata_writer,
+            self.my_member_id.storage_id,
+            self.status_tx,
+        );
+
+        Ok(passive)
+    }
+
+    fn should_leave(&self) -> bool {
+        let nodes_config = Metadata::with_current(|m| m.nodes_config_ref());
+        let my_node_config = nodes_config
+            .find_node_by_id(self.my_member_id.node_id)
+            .expect("my node config should be present");
+        // we use the NodesConfiguration as the authoritative signal for leaving the cluster
+        let should_leave = my_node_config.metadata_server_config.metadata_server_state
+            == MetadataServerState::Outsider;
```

Comment:
I realized that this won't reliably work in the case where the user removed the node and then shortly after added it back. If the node missed the actual configuration change and then reaches this point, it looks at the `MetadataServerStatus` and thinks that everything is ok.

Reply:
One idea for solving this problem could be to also introduce a transitional state for joining a cluster (similar to `Leaving`). That way, adding a node won't immediately set the state to `Member` and would tell a node which missed the configuration change to step down and rejoin.